vllm/core/scheduler.py

```diff
@@ -1202,7 +1217,8 @@ def _append_slots(
                 the new source and destination block indices for the appended
                 slots.
         """
-        num_lookahead_slots = self._get_num_lookahead_slots(is_prefill=False)
+        num_lookahead_slots = self._get_num_lookahead_slots(
+            is_prefill=seq_group.is_prefill())
```

Comment:
Spent some time looking into PP. The assertion issue may be related to this change. Something is causing the batch to generate logits for the prefills in the batch and since `len(sample_indices) == 0) as they don't perform sampling, the assertion fails on `assert logits_applied == logits.shape[0]` in `_apply_min_tokens_penalty` since logits_applied is the sum of all `sampled_indices` in the batch

Reply:
I see. Thanks @SolitaryThinker. I believe it should be 
```
        num_lookahead_slots = self._get_num_lookahead_slots(is_prefill=seq_group.is_prefill() and seq_group.do_sample)
```
instead of just 
```
        num_lookahead_slots = self._get_num_lookahead_slots(is_prefill=False)
```
I haven't been able to reproduce this assertion yet. I'll keep at it. When you find some time, can you try this out as well ? Thanks.

Reply:
https://github.com/vllm-project/vllm/pull/7814/commits/976d03275dd7a5f0d125769484dbe6a42a264c64 fixes the PP issue. The SamplingMetadata objects were being clobbered due to the SamplingMetadataCache reset.